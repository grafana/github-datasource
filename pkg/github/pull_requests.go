package github

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/grafana/github-datasource/pkg/models"
	"github.com/grafana/grafana-plugin-sdk-go/data"
	"github.com/pkg/errors"
	"github.com/shurcooL/githubv4"
)

// QueryListPullRequests lists all pull requests in a repository
// {
//   search(query: "is:pr repo:grafana/grafana merged:2020-08-19..*", type: ISSUE, first: 100) {
//     nodes {
//       ... on PullRequest {
//         id
//         title
//       }
//   }
// }
type QueryListPullRequests struct {
	Search struct {
		Nodes []struct {
			PullRequest PullRequest `graphql:"... on PullRequest"`
		}
		PageInfo PageInfo
	} `graphql:"search(query: $query, type: ISSUE, first: 100, after: $cursor)"`
}

// PullRequestAuthor is the structure of the Author object in a Pull Request (which requires a grapql object expansion on `User`)
type PullRequestAuthor struct {
	User User `graphql:"... on User"`
}

// PullRequest is a GitHub pull request
type PullRequest struct {
	Number     int64
	Title      string
	URL        string
	State      githubv4.PullRequestState
	Author     PullRequestAuthor
	Closed     bool
	IsDraft    bool
	Locked     bool
	Merged     bool
	ClosedAt   githubv4.DateTime
	CreatedAt  githubv4.DateTime
	UpdatedAt  githubv4.DateTime
	MergedAt   githubv4.DateTime
	Mergeable  githubv4.MergeableState
	MergedBy   *PullRequestAuthor
	Repository Repository
}

// PullRequests is a list of GitHub Pull Requests
type PullRequests []PullRequest

// Frames converts the list of Pull Requests to a Grafana DataFrame
func (p PullRequests) Frames() data.Frames {
	openTime := data.NewField("open_time", nil, []float64{})
	openTime.Config = &data.FieldConfig{
		Unit: "s", // The values are in seconds
	}

	frame := data.NewFrame(
		"pull_requests",
		data.NewField("number", nil, []int64{}),
		data.NewField("title", nil, []string{}),
		data.NewField("url", nil, []string{}),
		data.NewField("repository", nil, []string{}),
		data.NewField("state", nil, []string{}),
		data.NewField("author_login", nil, []string{}),
		data.NewField("author_email", nil, []string{}),
		data.NewField("author_company", nil, []string{}),
		data.NewField("closed", nil, []bool{}),
		data.NewField("is_draft", nil, []bool{}),
		data.NewField("locked", nil, []bool{}),
		data.NewField("merged", nil, []bool{}),
		data.NewField("mergeable", nil, []string{}),
		data.NewField("closed_at", nil, []*time.Time{}),
		data.NewField("merged_at", nil, []*time.Time{}),
		data.NewField("updated_at", nil, []time.Time{}),
		data.NewField("created_at", nil, []time.Time{}),
		openTime,
	)

	for _, v := range p {
		var (
			closedAt    *time.Time
			mergedAt    *time.Time
			secondsOpen float64 = time.Now().UTC().Sub(v.CreatedAt.UTC()).Round(time.Second).Seconds()
		)

		if !v.ClosedAt.IsZero() {
			t := v.ClosedAt.Time
			closedAt = &t
		}

		if !v.MergedAt.IsZero() {
			t := v.MergedAt.Time
			mergedAt = &t
		}

		if closedAt != nil {
			secondsOpen = v.ClosedAt.UTC().Sub(v.CreatedAt.UTC()).Seconds()
		}

		if mergedAt != nil {
			secondsOpen = v.ClosedAt.UTC().Sub(v.MergedAt.UTC()).Seconds()
		}

		frame.AppendRow(
			v.Number,
			v.Title,
			v.URL,
			v.Repository.NameWithOwner,
			string(v.State),
			v.Author.User.Login,
			v.Author.User.Email,
			v.Author.User.Company,
			v.Closed,
			v.IsDraft,
			v.Locked,
			v.Merged,
			string(v.Mergeable),
			closedAt,
			mergedAt,
			v.UpdatedAt.Time,
			v.CreatedAt.Time,
			secondsOpen,
		)
	}

	return data.Frames{frame}
}

// GetAllPullRequests uses the graphql search endpoint API to search all pull requests in the repository
func GetAllPullRequests(ctx context.Context, client Client, opts models.ListPullRequestsOptions) (PullRequests, error) {
	var (
		variables = map[string]interface{}{
			"cursor": (*githubv4.String)(nil),
			"query":  githubv4.String(buildQuery(opts)),
		}

		pullRequests = []PullRequest{}
	)

	for {
		q := &QueryListPullRequests{}
		if err := client.Query(ctx, q, variables); err != nil {
			return nil, errors.WithStack(err)
		}
		prs := make([]PullRequest, len(q.Search.Nodes))

		for i, v := range q.Search.Nodes {
			prs[i] = v.PullRequest
		}

		pullRequests = append(pullRequests, prs...)

		if !q.Search.PageInfo.HasNextPage {
			break
		}
		variables["cursor"] = q.Search.PageInfo.EndCursor
	}

	return pullRequests, nil
}

// GetPullRequestsInRange uses the graphql search endpoint API to find pull requests in the given time range.
func GetPullRequestsInRange(ctx context.Context, client Client, opts models.ListPullRequestsOptions, from time.Time, to time.Time) (PullRequests, error) {
	var q string

	if opts.TimeField != models.PullRequestNone {
		q = fmt.Sprintf("%s:%s..%s", opts.TimeField.String(), from.Format(time.RFC3339), to.Format(time.RFC3339))
	}

	if opts.Query != nil {
		q = fmt.Sprintf("%s %s", *opts.Query, q)
	}

	return GetAllPullRequests(ctx, client, models.ListPullRequestsOptions{
		Repository: opts.Repository,
		Owner:      opts.Owner,
		TimeField:  opts.TimeField,
		Query:      &q,
	})
}

// buildQuery builds the "query" field for Pull Request searches
func buildQuery(opts models.ListPullRequestsOptions) string {
	search := []string{
		"is:pr",
	}

	if opts.Repository == "" {
		search = append(search, fmt.Sprintf("org:%s", opts.Owner))
	} else {
		search = append(search, fmt.Sprintf("repo:%s/%s", opts.Owner, opts.Repository))
	}

	if opts.Query != nil {
		search = append(search, *opts.Query)
	}

	return strings.Join(search, " ")
}
