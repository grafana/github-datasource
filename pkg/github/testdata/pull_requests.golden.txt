ðŸŒŸ This was machine generated.  Do not edit. ðŸŒŸ

Frame[0] 
Name: pull_requests
Dimensions: 18 Fields by 3 Rows
+---------------+----------------+------------------------------------------------------+---------------------------+----------------+--------------------+--------------------+----------------------+--------------+----------------+--------------+--------------+-----------------+---------------------------------+---------------------------------+---------------------------------+---------------------------------+------------------------+
| Name: number  | Name: title    | Name: url                                            | Name: repository          | Name: state    | Name: author_login | Name: author_email | Name: author_company | Name: closed | Name: is_draft | Name: locked | Name: merged | Name: mergeable | Name: closed_at                 | Name: merged_at                 | Name: updated_at                | Name: created_at                | Name: open_time        |
| Labels:       | Labels:        | Labels:                                              | Labels:                   | Labels:        | Labels:            | Labels:            | Labels:              | Labels:      | Labels:        | Labels:      | Labels:      | Labels:         | Labels:                         | Labels:                         | Labels:                         | Labels:                         | Labels:                |
| Type: []int64 | Type: []string | Type: []string                                       | Type: []string            | Type: []string | Type: []string     | Type: []string     | Type: []string       | Type: []bool | Type: []bool   | Type: []bool | Type: []bool | Type: []string  | Type: []*time.Time              | Type: []*time.Time              | Type: []time.Time               | Type: []time.Time               | Type: []float64        |
+---------------+----------------+------------------------------------------------------+---------------------------+----------------+--------------------+--------------------+----------------------+--------------+----------------+--------------+--------------+-----------------+---------------------------------+---------------------------------+---------------------------------+---------------------------------+------------------------+
| 1             | PullRequest #1 | https://github.com/grafana/github-datasource/pulls/1 | grafana/github-datasource | OPEN           | testUser           | user@example.com   | ACME corp            | true         | false          | false        | true         | MERGEABLE       | 2020-08-25 14:41:56 +0000 +0000 | 2020-08-25 14:41:56 +0000 +0000 | 2020-08-25 16:21:56 +0000 +0000 | 2020-08-25 16:21:56 +0000 +0000 | 0                      |
| 2             | PullRequest #2 | https://github.com/grafana/github-datasource/pulls/2 | grafana/github-datasource | OPEN           | testUser2          | user2@example.com  | ACME corp            | true         | false          | false        | true         | MERGEABLE       | 2020-08-25 14:41:56 +0000 +0000 | 2020-08-25 14:41:56 +0000 +0000 | 2020-08-25 18:21:56 +0000 +0000 | 2020-08-25 16:21:56 +0000 +0000 | 0                      |
| 3             | PullRequest #2 | https://github.com/grafana/github-datasource/pulls/3 | grafana/github-datasource | OPEN           | testUser2          | user2@example.com  | ACME corp            | false        | false          | false        | false        | MERGEABLE       | null                            | 2020-08-25 14:41:56 +0000 +0000 | 2020-08-25 18:21:56 +0000 +0000 | 2020-08-25 16:21:56 +0000 +0000 | -9.223372036854776e+09 |
+---------------+----------------+------------------------------------------------------+---------------------------+----------------+--------------------+--------------------+----------------------+--------------+----------------+--------------+--------------+-----------------+---------------------------------+---------------------------------+---------------------------------+---------------------------------+------------------------+


====== TEST DATA RESPONSE (arrow base64) ======
FRAME=QVJST1cxAAD/////+AcAABAAAAAAAAoADgAMAAsABAAKAAAAFAAAAAAAAAEDAAoADAAAAAgABAAKAAAACAAAAFwAAAACAAAAKAAAAAQAAACY+P//CAAAAAwAAAAAAAAAAAAAAAUAAAByZWZJZAAAALj4//8IAAAAGAAAAA0AAABwdWxsX3JlcXVlc3RzAAAABAAAAG5hbWUAAAAAEgAAAPgGAACIBgAANAYAANAFAAB0BQAACAUAAJwEAAAwBAAA1AMAAHADAAAUAwAAuAIAAFQCAADkAQAAbAEAAAQBAACcAAAABAAAAGL5//8UAAAAcAAAAHAAAAAAAAADcAAAAAIAAAAwAAAABAAAAFT5//8IAAAAFAAAAAkAAABvcGVuX3RpbWUAAAAEAAAAbmFtZQAAAAB8+f//CAAAABgAAAAMAAAAeyJ1bml0IjoicyJ9AAAAAAYAAABjb25maWcAAAAAAABe/v//AAACAAkAAABvcGVuX3RpbWUAAAD2+f//FAAAAEAAAABAAAAAAAAACkAAAAABAAAABAAAAOT5//8IAAAAFAAAAAoAAABjcmVhdGVkX2F0AAAEAAAAbmFtZQAAAAAAAAAAwv7//wAAAwAKAAAAY3JlYXRlZF9hdAAAWvr//xQAAABAAAAAQAAAAAAAAApAAAAAAQAAAAQAAABI+v//CAAAABQAAAAKAAAAdXBkYXRlZF9hdAAABAAAAG5hbWUAAAAAAAAAACb///8AAAMACgAAAHVwZGF0ZWRfYXQAAJ7///8UAAAAQAAAAEAAAAAAAAoBQAAAAAEAAAAEAAAArPr//wgAAAAUAAAACQAAAG1lcmdlZF9hdAAAAAQAAABuYW1lAAAAAAAAAACK////AAADAAkAAABtZXJnZWRfYXQAEgAYABQAEwASAAwAAAAIAAQAEgAAABQAAABAAAAASAAAAAAACgFIAAAAAQAAAAQAAAAg+///CAAAABQAAAAJAAAAY2xvc2VkX2F0AAAABAAAAG5hbWUAAAAAAAAAAAAABgAIAAYABgAAAAAAAwAJAAAAY2xvc2VkX2F0AAAAnvv//xQAAABAAAAAQAAAAAAAAAU8AAAAAQAAAAQAAACM+///CAAAABQAAAAJAAAAbWVyZ2VhYmxlAAAABAAAAG5hbWUAAAAAAAAAAPz7//8JAAAAbWVyZ2VhYmxlAAAA/vv//xQAAAA8AAAAPAAAAAAAAAY4AAAAAQAAAAQAAADs+///CAAAABAAAAAGAAAAbWVyZ2VkAAAEAAAAbmFtZQAAAAAAAAAAWPz//wYAAABtZXJnZWQAAFb8//8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAARPz//wgAAAAQAAAABgAAAGxvY2tlZAAABAAAAG5hbWUAAAAAAAAAALD8//8GAAAAbG9ja2VkAACu/P//FAAAAEAAAABAAAAAAAAABjwAAAABAAAABAAAAJz8//8IAAAAFAAAAAgAAABpc19kcmFmdAAAAAAEAAAAbmFtZQAAAAAAAAAADP3//wgAAABpc19kcmFmdAAAAAAO/f//FAAAADwAAAA8AAAAAAAABjgAAAABAAAABAAAAPz8//8IAAAAEAAAAAYAAABjbG9zZWQAAAQAAABuYW1lAAAAAAAAAABo/f//BgAAAGNsb3NlZAAAZv3//xQAAABEAAAARAAAAAAAAAVAAAAAAQAAAAQAAABU/f//CAAAABgAAAAOAAAAYXV0aG9yX2NvbXBhbnkAAAQAAABuYW1lAAAAAAAAAADI/f//DgAAAGF1dGhvcl9jb21wYW55AADO/f//FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAALz9//8IAAAAGAAAAAwAAABhdXRob3JfZW1haWwAAAAABAAAAG5hbWUAAAAAAAAAADD+//8MAAAAYXV0aG9yX2VtYWlsAAAAADb+//8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAAJP7//wgAAAAYAAAADAAAAGF1dGhvcl9sb2dpbgAAAAAEAAAAbmFtZQAAAAAAAAAAmP7//wwAAABhdXRob3JfbG9naW4AAAAAnv7//xQAAAA8AAAAPAAAAAAAAAU4AAAAAQAAAAQAAACM/v//CAAAABAAAAAFAAAAc3RhdGUAAAAEAAAAbmFtZQAAAAAAAAAA+P7//wUAAABzdGF0ZQAAAPb+//8UAAAAQAAAAEAAAAAAAAAFPAAAAAEAAAAEAAAA5P7//wgAAAAUAAAACgAAAHJlcG9zaXRvcnkAAAQAAABuYW1lAAAAAAAAAABU////CgAAAHJlcG9zaXRvcnkAAFb///8UAAAAOAAAADgAAAAAAAAFNAAAAAEAAAAEAAAARP///wgAAAAMAAAAAwAAAHVybAAEAAAAbmFtZQAAAAAAAAAArP///wMAAAB1cmwApv///xQAAAA8AAAAQAAAAAAAAAU8AAAAAQAAAAQAAACU////CAAAABAAAAAFAAAAdGl0bGUAAAAEAAAAbmFtZQAAAAAAAAAABAAEAAQAAAAFAAAAdGl0bGUAEgAYABQAAAATAAwAAAAIAAQAEgAAABQAAABEAAAATAAAAAAAAAJQAAAAAQAAAAwAAAAIAAwACAAEAAgAAAAIAAAAEAAAAAYAAABudW1iZXIAAAQAAABuYW1lAAAAAAAAAAAIAAwACAAHAAgAAAAAAAABQAAAAAYAAABudW1iZXIAAP////84BAAAFAAAAAAAAAAMABYAFAATAAwABAAMAAAAAAMAAAAAAAAUAAAAAAAAAwMACgAYAAwACAAEAAoAAAAUAAAA2AIAAAMAAAAAAAAAAAAAACwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYAAAAAAAAABgAAAAAAAAAAAAAAAAAAAAYAAAAAAAAABAAAAAAAAAAKAAAAAAAAAAwAAAAAAAAAFgAAAAAAAAAAAAAAAAAAABYAAAAAAAAABAAAAAAAAAAaAAAAAAAAACgAAAAAAAAAAgBAAAAAAAAAAAAAAAAAAAIAQAAAAAAABAAAAAAAAAAGAEAAAAAAABQAAAAAAAAAGgBAAAAAAAAAAAAAAAAAABoAQAAAAAAABAAAAAAAAAAeAEAAAAAAAAQAAAAAAAAAIgBAAAAAAAAAAAAAAAAAACIAQAAAAAAABAAAAAAAAAAmAEAAAAAAAAgAAAAAAAAALgBAAAAAAAAAAAAAAAAAAC4AQAAAAAAABAAAAAAAAAAyAEAAAAAAAA4AAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAgAAAAAAABAAAAAAAAAAEAIAAAAAAAAgAAAAAAAAADACAAAAAAAAAAAAAAAAAAAwAgAAAAAAAAgAAAAAAAAAOAIAAAAAAAAAAAAAAAAAADgCAAAAAAAACAAAAAAAAABAAgAAAAAAAAAAAAAAAAAAQAIAAAAAAAAIAAAAAAAAAEgCAAAAAAAAAAAAAAAAAABIAgAAAAAAAAgAAAAAAAAAUAIAAAAAAAAAAAAAAAAAAFACAAAAAAAAEAAAAAAAAABgAgAAAAAAACAAAAAAAAAAgAIAAAAAAAAIAAAAAAAAAIgCAAAAAAAAGAAAAAAAAACgAgAAAAAAAAAAAAAAAAAAoAIAAAAAAAAYAAAAAAAAALgCAAAAAAAAAAAAAAAAAAC4AgAAAAAAABgAAAAAAAAA0AIAAAAAAAAAAAAAAAAAANACAAAAAAAAGAAAAAAAAADoAgAAAAAAAAAAAAAAAAAA6AIAAAAAAAAYAAAAAAAAAAAAAAASAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAQAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAACAAAAAAAAAAMAAAAAAAAAAAAAAA4AAAAcAAAAKgAAAFB1bGxSZXF1ZXN0ICMxUHVsbFJlcXVlc3QgIzJQdWxsUmVxdWVzdCAjMgAAAAAAAAAAAAA0AAAAaAAAAJwAAABodHRwczovL2dpdGh1Yi5jb20vZ3JhZmFuYS9naXRodWItZGF0YXNvdXJjZS9wdWxscy8xaHR0cHM6Ly9naXRodWIuY29tL2dyYWZhbmEvZ2l0aHViLWRhdGFzb3VyY2UvcHVsbHMvMmh0dHBzOi8vZ2l0aHViLmNvbS9ncmFmYW5hL2dpdGh1Yi1kYXRhc291cmNlL3B1bGxzLzMAAAAAAAAAABkAAAAyAAAASwAAAGdyYWZhbmEvZ2l0aHViLWRhdGFzb3VyY2VncmFmYW5hL2dpdGh1Yi1kYXRhc291cmNlZ3JhZmFuYS9naXRodWItZGF0YXNvdXJjZQAAAAAAAAAAAAQAAAAIAAAADAAAAE9QRU5PUEVOT1BFTgAAAAAAAAAACAAAABEAAAAaAAAAdGVzdFVzZXJ0ZXN0VXNlcjJ0ZXN0VXNlcjIAAAAAAAAAAAAAEAAAACEAAAAyAAAAdXNlckBleGFtcGxlLmNvbXVzZXIyQGV4YW1wbGUuY29tdXNlcjJAZXhhbXBsZS5jb20AAAAAAAAAAAAACQAAABIAAAAbAAAAQUNNRSBjb3JwQUNNRSBjb3JwQUNNRSBjb3JwAAAAAAADAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAAAAAAAAAAAAAAAJAAAAEgAAABsAAABNRVJHRUFCTEVNRVJHRUFCTEVNRVJHRUFCTEUAAAAAAAMAAAAAAAAAAAgPt+CJLhYACA+34IkuFgAAAAAAAAAAAAgPt+CJLhYACA+34IkuFgAID7fgiS4WAGjtslWPLhYAqF4U4pUuFgCoXhTilS4WAGjtslWPLhYAaO2yVY8uFgBo7bJVjy4WAAAAAAAAAAAAAAAAAAAAAJXWJugLLgHCEAAAAAwAFAASAAwACAAEAAwAAAAQAAAALAAAADwAAAAAAAMAAQAAAAgIAAAAAAAAQAQAAAAAAAAAAwAAAAAAAAAAAAAAAAAAAAAAAAAACgAMAAAACAAEAAoAAAAIAAAAXAAAAAIAAAAoAAAABAAAAJj4//8IAAAADAAAAAAAAAAAAAAABQAAAHJlZklkAAAAuPj//wgAAAAYAAAADQAAAHB1bGxfcmVxdWVzdHMAAAAEAAAAbmFtZQAAAAASAAAA+AYAAIgGAAA0BgAA0AUAAHQFAAAIBQAAnAQAADAEAADUAwAAcAMAABQDAAC4AgAAVAIAAOQBAABsAQAABAEAAJwAAAAEAAAAYvn//xQAAABwAAAAcAAAAAAAAANwAAAAAgAAADAAAAAEAAAAVPn//wgAAAAUAAAACQAAAG9wZW5fdGltZQAAAAQAAABuYW1lAAAAAHz5//8IAAAAGAAAAAwAAAB7InVuaXQiOiJzIn0AAAAABgAAAGNvbmZpZwAAAAAAAF7+//8AAAIACQAAAG9wZW5fdGltZQAAAPb5//8UAAAAQAAAAEAAAAAAAAAKQAAAAAEAAAAEAAAA5Pn//wgAAAAUAAAACgAAAGNyZWF0ZWRfYXQAAAQAAABuYW1lAAAAAAAAAADC/v//AAADAAoAAABjcmVhdGVkX2F0AABa+v//FAAAAEAAAABAAAAAAAAACkAAAAABAAAABAAAAEj6//8IAAAAFAAAAAoAAAB1cGRhdGVkX2F0AAAEAAAAbmFtZQAAAAAAAAAAJv///wAAAwAKAAAAdXBkYXRlZF9hdAAAnv///xQAAABAAAAAQAAAAAAACgFAAAAAAQAAAAQAAACs+v//CAAAABQAAAAJAAAAbWVyZ2VkX2F0AAAABAAAAG5hbWUAAAAAAAAAAIr///8AAAMACQAAAG1lcmdlZF9hdAASABgAFAATABIADAAAAAgABAASAAAAFAAAAEAAAABIAAAAAAAKAUgAAAABAAAABAAAACD7//8IAAAAFAAAAAkAAABjbG9zZWRfYXQAAAAEAAAAbmFtZQAAAAAAAAAAAAAGAAgABgAGAAAAAAADAAkAAABjbG9zZWRfYXQAAACe+///FAAAAEAAAABAAAAAAAAABTwAAAABAAAABAAAAIz7//8IAAAAFAAAAAkAAABtZXJnZWFibGUAAAAEAAAAbmFtZQAAAAAAAAAA/Pv//wkAAABtZXJnZWFibGUAAAD++///FAAAADwAAAA8AAAAAAAABjgAAAABAAAABAAAAOz7//8IAAAAEAAAAAYAAABtZXJnZWQAAAQAAABuYW1lAAAAAAAAAABY/P//BgAAAG1lcmdlZAAAVvz//xQAAAA8AAAAPAAAAAAAAAY4AAAAAQAAAAQAAABE/P//CAAAABAAAAAGAAAAbG9ja2VkAAAEAAAAbmFtZQAAAAAAAAAAsPz//wYAAABsb2NrZWQAAK78//8UAAAAQAAAAEAAAAAAAAAGPAAAAAEAAAAEAAAAnPz//wgAAAAUAAAACAAAAGlzX2RyYWZ0AAAAAAQAAABuYW1lAAAAAAAAAAAM/f//CAAAAGlzX2RyYWZ0AAAAAA79//8UAAAAPAAAADwAAAAAAAAGOAAAAAEAAAAEAAAA/Pz//wgAAAAQAAAABgAAAGNsb3NlZAAABAAAAG5hbWUAAAAAAAAAAGj9//8GAAAAY2xvc2VkAABm/f//FAAAAEQAAABEAAAAAAAABUAAAAABAAAABAAAAFT9//8IAAAAGAAAAA4AAABhdXRob3JfY29tcGFueQAABAAAAG5hbWUAAAAAAAAAAMj9//8OAAAAYXV0aG9yX2NvbXBhbnkAAM79//8UAAAARAAAAEQAAAAAAAAFQAAAAAEAAAAEAAAAvP3//wgAAAAYAAAADAAAAGF1dGhvcl9lbWFpbAAAAAAEAAAAbmFtZQAAAAAAAAAAMP7//wwAAABhdXRob3JfZW1haWwAAAAANv7//xQAAABEAAAARAAAAAAAAAVAAAAAAQAAAAQAAAAk/v//CAAAABgAAAAMAAAAYXV0aG9yX2xvZ2luAAAAAAQAAABuYW1lAAAAAAAAAACY/v//DAAAAGF1dGhvcl9sb2dpbgAAAACe/v//FAAAADwAAAA8AAAAAAAABTgAAAABAAAABAAAAIz+//8IAAAAEAAAAAUAAABzdGF0ZQAAAAQAAABuYW1lAAAAAAAAAAD4/v//BQAAAHN0YXRlAAAA9v7//xQAAABAAAAAQAAAAAAAAAU8AAAAAQAAAAQAAADk/v//CAAAABQAAAAKAAAAcmVwb3NpdG9yeQAABAAAAG5hbWUAAAAAAAAAAFT///8KAAAAcmVwb3NpdG9yeQAAVv///xQAAAA4AAAAOAAAAAAAAAU0AAAAAQAAAAQAAABE////CAAAAAwAAAADAAAAdXJsAAQAAABuYW1lAAAAAAAAAACs////AwAAAHVybACm////FAAAADwAAABAAAAAAAAABTwAAAABAAAABAAAAJT///8IAAAAEAAAAAUAAAB0aXRsZQAAAAQAAABuYW1lAAAAAAAAAAAEAAQABAAAAAUAAAB0aXRsZQASABgAFAAAABMADAAAAAgABAASAAAAFAAAAEQAAABMAAAAAAAAAlAAAAABAAAADAAAAAgADAAIAAQACAAAAAgAAAAQAAAABgAAAG51bWJlcgAABAAAAG5hbWUAAAAAAAAAAAgADAAIAAcACAAAAAAAAAFAAAAABgAAAG51bWJlcgAAKAgAAEFSUk9XMQ==
